src/osd/pg_scrubber.h

```diff
@@ -0,0 +1,677 @@
+  // the I/F used by the state-machine (i.e. the implementation of ScrubMachineListener)
+
+  bool select_range() final;
+
+  /// walk the log to find the latest update that affects our chunk
+  eversion_t search_log_for_updates() const final;
+
+  eversion_t get_last_update_applied() const final
+  {
+    return m_pg->recovery_state.get_last_update_applied();
+  }
+
+  void requeue_waiting() const final { m_pg->requeue_ops(m_pg->waiting_for_scrub); }
+
+  int pending_active_pushes() const final { return m_pg->active_pushes; }
+
+  void scrub_compare_maps() final;
+
+  void on_init() final;
+  void on_replica_init() final;
+  void replica_handling_done() final;
+
+  /// the version of 'scrub_clear_state()' that does not try to invoke FSM services
+  /// (thus can be called from FSM reactions)
+  void clear_pgscrub_state(bool keep_repair_state) final;
+
+  void add_delayed_scheduling() final;
+
+  /**
+   * @returns have we asked at least one replica?
+   * 'false' means we are configured with no replicas, and
+   * should expect no maps to arrive.
+   */
+  bool get_replicas_maps(bool replica_can_preempt) final;
+
+  Scrub::FsmNext on_digest_updates() final;
+
+  void send_replica_map(bool was_preempted) final;
+
+  void send_remotes_reserved() final;
+  void send_reservation_failure() final;
+
+  /**
+   *  does the PG have newer updates than what we (the scrubber) know?
+   */
+  [[nodiscard]] bool has_pg_marked_new_updates() const final;
+
+  void set_subset_last_update(eversion_t e) final;
+
+  void replica_update_start_epoch() final;
+
+  void maps_compare_n_cleanup() final;
+
+  Scrub::preemption_t* get_preemptor() final;
+
+  int build_primary_map_chunk() final;
+
+  int build_replica_map_chunk() final;
+
+  void reserve_replicas() final;
+
+  [[nodiscard]] bool was_epoch_changed() const final;
+
+  void mark_local_map_ready() final;
+
+  [[nodiscard]] bool are_all_maps_available() const final;
+
+  std::string dump_awaited_maps() const final;
+
+ protected:
+  bool state_test(uint64_t m) const { return m_pg->state_test(m); }
+  void state_set(uint64_t m) { m_pg->state_set(m); }
+  void state_clear(uint64_t m) { m_pg->state_clear(m); }
+
+  [[nodiscard]] bool is_primary() const { return m_pg->recovery_state.is_primary(); }
+
+  [[nodiscard]] bool is_scrub_registered() const;
+
+  virtual void _scrub_clear_state() {}
+
+  utime_t m_scrub_reg_stamp;  ///< stamp we registered for
+
+  ostream& show(ostream& out) const override;
+
+ public:
+  // -------------------------------------------------------------------------------------------
+
+  friend ostream& operator<<(ostream& out, const PgScrubber& scrubber);
+
+  static utime_t scrub_must_stamp() { return utime_t(1, 1); }
+
+  virtual ~PgScrubber();  // must be defined separately, in the .cc file
+
+  [[nodiscard]] bool is_scrub_active() const final;
+
+ private:
+  void reset_internal_state();
+
+  void _scan_snaps(ScrubMap& smap);  // note that the (non-standard for a
```

Comment:
By this, do you mean that it looks for a log line with "_scan_snaps" in it?  Surely that simply means that modifying that log line means modifying the test.  Maybe just stick a comment next to the log line?

Reply:
modified